### PR TITLE
Feature/web3 react(web3.js)

### PR DESCRIPTION
**Pull request by aliangm**
[x] Before opening a pull request, please read the [contributing guidelines](https://github.com/pancakeswap/pancake-frontend/blob/master/CONTRIBUTING.md) first
[x] If your PR is work in progress, open it as `draft`
[x] Before requesting a review, all the checks need to pass
[x] Explain what your PR does

- Upgrade @binance-chain/bsc-use-wallet to @web3-react package

### CODE REVIEW

File: .env.development [hunk 1]
```diff
@@ -2,7 +2,7 @@ REACT_APP_CHAIN_ID = "56"
 REACT_APP_GTAG = "GTM-PXLD3XW"
 
 # 10+ nodes balanced, US/EU
-REACT_APP_NODE_1 = "https://bsc-dataseed1.ninicoin.io"
+REACT_APP_NODE_1 = "https://bsc-dataseed1.binance.io"
```
Comment (Chef-Chungus) [inline .env.development:5]: Why did you change the node?

File: package.json [hunk 2]
```diff
@@ -31,24 +31,30 @@
...
     "canvas-confetti": "^1.3.3",
     "chart.js": "^2.9.4",
     "date-fns": "^2.17.0",
+    "ethers": "^5.0.31",
     "lodash": "^4.17.20",
     "node-fetch": "^2.6.1",
     "react": "^17.0.1",
     "react-chartjs-2": "^2.11.1",
     "react-countup": "^4.3.3",
+    "react-device-detect": "^1.15.0",
```
Comment (hachiojidev) [inline package.json:57]: Why is this needed?

File: src/connectors/index.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+export const network = new NetworkConnector({
+  urls: { [NETWORK_CHAIN_ID]: NETWORK_URL }
+})
+
+let networkLibrary: Web3Provider | undefined
+export async function getNetworkLibrary(): Promise<Web3Provider> {
+  const provider = await network.getProvider()
+  networkLibrary = networkLibrary ?? new Web3Provider(provider as any)
+  return networkLibrary
+}
```
Comment (hachiojidev) [inline src/connectors/index.ts:25]: Can you explain a bit why `networkLibrary` is necessary outside the function?

File: src/components/Web3ReactManager/index.tsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import React, { useState, useEffect } from 'react'
+import { useWeb3React } from '@web3-react/core'
+
+import useEagerConnect from 'hooks/useEagerConnect'
+import useInactiveListener from 'hooks/useInactiveListener'
+import PageLoader from 'components/PageLoader'
+import { network } from 'connectors'
+
+export default function Web3ReactManager({ children }: { children: JSX.Element }) {
```
Comment (hachiojidev) [inline src/components/Web3ReactManager/index.tsx:9]: This does not follow our pattern with typing or functions.

File: src/components/Web3ReactManager/index.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  const [showLoader, setShowLoader] = useState(false)
+  useEffect(() => {
+    const timeout = setTimeout(() => {
+      setShowLoader(true)
+    }, 600)
+
+    return () => {
+      clearTimeout(timeout)
+    }
+  }, [])
```
Comment (hachiojidev) [inline src/components/Web3ReactManager/index.tsx:35]: Why are we delaying the loading state?

File: src/components/Web3ReactManager/index.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  // if the account context isn't active, and there's an error on the network context, it's an irrecoverable error
+  if (!active && error) {
+    // return (
+    //   <MessageWrapper>
+    //     <Message>{t('unknownError')}</Message>
+    //   </MessageWrapper>
+    // )
+  }
```
Comment (hachiojidev) [inline src/components/Web3ReactManager/index.tsx:49]: I hope we're going to do something here.

File: src/components/Web3ReactManager/index.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+    return () => {
+      clearTimeout(timeout)
+    }
+  }, [])
+
+  // on page load, do nothing until we've tried to connect to the injected connector
+  if (!triedEager) {
+    return null
+  }
```
Comment (hachiojidev) [inline src/components/Web3ReactManager/index.tsx:40]: I am not sure this is the best approach. The user will see a blank white page.

**Review by hachiojidev — COMMENTED** [review]
When I click on Binance Smart Wallet and don't have it on my browser the site reverts to a blank page.

File: src/components/UnlockButton.tsx [hunk 8]
```diff
@@ -1,12 +1,26 @@
...
 import useI18n from 'hooks/useI18n'
 
 const UnlockButton = (props) => {
   const TranslateString = useI18n()
-  const { connect, reset } = useWallet()
-  const { onPresentConnectModal } = useWalletModal(connect, reset)
+  const { account, activate, deactivate, connector } = useWeb3React()
+  const handleLogin = (connectorId: ConnectorId) => {
+    if (connectorId === 'walletconnect') {
+      if (connector instanceof WalletConnectConnector && connector.walletConnectProvider?.wc?.uri) {
```
Comment (RabbitDoge) [inline src/components/UnlockButton.tsx:14]: What this if block is for ?

File: src/App.tsx [hunk 9]
```diff
@@ -41,9 +42,9 @@ const App: React.FC = () => {
 
   useEffect(() => {
```
Comment (RabbitDoge) [inline src/App.tsx:43]: isn't it redundant with `useEagerConnect` ?

File: src/components/Web3ReactManager/index.tsx [hunk 10]
```diff
@@ -0,0 +1,59 @@
+import React, { useState, useEffect } from 'react'
```
Comment (RabbitDoge) [inline src/components/Web3ReactManager/index.tsx:1]: I don't know what you guys think (@aliangm @hachiojidev ) but for me , this component and the `useEagerConnect` should be merged together. The logic in this component seems to be a continuation of the `useEagerConnect` hook, and I don't see a use case where we only want to use `useEagerConnect`.

Also, it seems that this component render a loaders while trying to connect, we probably want to render the website even if the user is not connected....

File: src/connectors/index.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+export const injected = new InjectedConnector({
+  supportedChainIds: [56, 97]
+})
+
+// mainnet only
+export const walletconnect = new WalletConnectConnector({
+  rpc: { 1: NETWORK_URL },
```
Comment (RabbitDoge) [inline src/connectors/index.ts:33]: Wild guess, but isn't `1` the network id ? It's not going to work.

File: src/connectors/index.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+let networkLibrary: Web3Provider | undefined
+export async function getNetworkLibrary(): Promise<Web3Provider> {
+  const provider = await network.getProvider()
+  networkLibrary = networkLibrary ?? new Web3Provider(provider as any)
+  return networkLibrary
+}
+
+export const injected = new InjectedConnector({
+  supportedChainIds: [56, 97]
```
Comment (RabbitDoge) [inline src/connectors/index.ts:28]: Supported chain ids should be taken from the env files, in case we support different setup with local blockchains later

File: .env.development (hunk 1 above)
Comment (aliangm) [reply .env.development:5]: I changed it because I'm unable to get response from ninicoin. I will revert it back

File: package.json (hunk 2 above)
Comment (aliangm) [reply package.json:57]: Removed

File: src/connectors/index.ts (hunk 3 above)
Comment (aliangm) [reply src/connectors/index.ts:25]: No need for now. I removed it

File: src/utils/getLibrary.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+import { Web3Provider } from '@ethersproject/providers'
```
Comment (RabbitDoge) [inline src/utils/getLibrary.ts:1]: This function should instantiate a web3 instance according to the doc https://github.com/NoahZinsmeister/web3-react/tree/v6/docs#props , not just returning the provider. 

If we do so, the `useWeb3` hook become obsolete, because we can retrieve the web3 instance from the `useWeb3React` hook

File: src/components/Web3ReactManager/index.tsx (hunk 4 above)
Comment (aliangm) [reply src/components/Web3ReactManager/index.tsx:9]: Updated

File: src/components/Web3ReactManager/index.tsx (hunk 5 above)
Comment (aliangm) [reply src/components/Web3ReactManager/index.tsx:35]: As it's Web3ReactManager is wrapper, it will be loaded before children. That's why I added delay

File: src/components/Web3ReactManager/index.tsx (hunk 6 above)
Comment (aliangm) [reply src/components/Web3ReactManager/index.tsx:49]: I was going to add toast

File: src/components/Web3ReactManager/index.tsx (hunk 7 above)
Comment (aliangm) [reply src/components/Web3ReactManager/index.tsx:40]: This case could be removed as we always try with injected provider when users unable to connect wallet

File: src/components/UnlockButton.tsx (hunk 8 above)
Comment (aliangm) [reply src/components/UnlockButton.tsx:14]: It's for walletconnect provider init issue

Comment (aliangm) [reply src/components/UnlockButton.tsx:14]: When users sign in again using walletconnect, it produces bug

File: src/App.tsx (hunk 9 above)
Comment (aliangm) [reply src/App.tsx:43]: Correct

File: src/connectors/index.ts (hunk 11 above)
Comment (aliangm) [reply src/connectors/index.ts:33]: Fixed

File: src/connectors/index.ts (hunk 12 above)
Comment (aliangm) [reply src/connectors/index.ts:28]: Fixed

File: src/utils/getLibrary.ts (hunk 13 above)
Comment (aliangm) [reply src/utils/getLibrary.ts:1]: I updated this part because if we change, we should update the part used `useWeb3` hook

File: src/components/Web3ReactManager/index.tsx (hunk 5 above)
Comment (hachiojidev) [reply src/components/Web3ReactManager/index.tsx:37]: What happens without the delay?

File: src/connectors/NetworkConnector.ts [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        if ('error' in result) {
+          reject(new RequestError(result?.error?.message, result?.error?.code, result?.error?.data))
+        } else if ('result' in result) {
+          resolve(result.result)
+        } else {
+          reject(new RequestError(`Received unexpected JSON-RPC response to ${method} request.`, -32000, result))
+        }
+      }
+    }
+  }
```
Comment (hachiojidev) [inline src/connectors/NetworkConnector.ts:112]: Is there a reason for this to be outside the `try` block? If not I think we should put it inside.

File: src/hooks/useInactiveListener.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
+import { useWeb3React as useWeb3ReactCore } from '@web3-react/core'
+import { useEffect } from 'react'
+import { injected } from '../connectors'
+
+export default function useInactiveListener(suppress = false) {
+  const { active, error, activate } = useWeb3ReactCore() // specifically using useWeb3React because of what this hook does
+
+  useEffect(() => {
+    const { ethereum } = window
```
Comment (hachiojidev) [inline src/hooks/useInactiveListener.ts:9]: Is working on the global `ethereum` object absolutely necessary? We want to avoid globals if we can.

File: src/react-app-env.d.ts [hunk 16]
```diff
@@ -1 +1,9 @@
 /// <reference types="react-scripts" />
+interface Window {
+  ethereum?: {
+    isMetaMask?: true
+    on?: (...args: any[]) => void
+    removeListener?: (...args: any[]) => void
+    autoRefreshOnNetworkChange?: boolean
+  }
+}
```
Comment (hachiojidev) [inline src/react-app-env.d.ts:9]: I am not sure but does this extend the `window` interface or overwrite it? It is the latter then we should make sure this extends.

File: src/connectors/NetworkConnector.ts (hunk 14 above)
Comment (self-coding-crab) [reply src/connectors/NetworkConnector.ts:112]: I used NetworkConnector from uniswap-interface coz they made web3.js compatible connector.
It seems we can put inside

File: src/hooks/useInactiveListener.ts (hunk 15 above)
Comment (self-coding-crab) [reply src/hooks/useInactiveListener.ts:9]: It's necessary for only inactive listener

File: src/react-app-env.d.ts (hunk 16 above)
Comment (self-coding-crab) [reply src/react-app-env.d.ts:9]: It does extend the interface I think but let me confirm

**Comment by RabbitDoge** [comment]
#528
